### PR TITLE
SALTO-2986 SF flow deployment

### PR DESCRIPTION
**Pull request by Uri-Bar**
_SF flow deployment_

---
_Additional context for reviewer_:
None

---
_Release Notes_: 
Salesforce-adapter:

* Add change validator to check all the different active flows deployment 
---

_User Notifications_: 
None

### CODE REVIEW

File: packages/salesforce-adapter/src/change_validators/active_flows.ts [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    }
+    const isActiveDeployEnabled = isUndefined(flowSetting)
+      ? false : flowSetting.value.enableFlowDeployAsActiveEnabled
+    const flowChanges = await awu(changes)
+      .filter(isInstanceChange)
+      .filter(isAdditionOrModificationChange)
+      .filter(isFlowChange)
+      .toArray()
+
+    const deactivateFlows = flowChanges
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/change_validators/active_flows.ts:197]: Nit:
```suggestion
    const deactivatingFlowChangeErrors = flowChanges
```

**Review by amitgoldarb — COMMENTED** [review]
This is looking good!
Added a few small comments, I know this is in draft so might be a bit too early.

File: packages/salesforce-adapter/src/change_validators/active_flows.ts [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    // TODO: change to preferActiveFlowVersions
+    const isPreferActiveVersion = config.enumFieldPermissions ?? false
+    if (!isUndefined(elementsSource)) {
+      flowSetting = await elementsSource.get(new ElemID(SALESFORCE, 'FlowSettings', 'instance'))
+    }
+    const isActiveDeployEnabled = isUndefined(flowSetting)
+      ? false : flowSetting.value.enableFlowDeployAsActiveEnabled
+    const flowChanges = await awu(changes)
+      .filter(isInstanceChange)
+      .filter(isAdditionOrModificationChange)
```
[the remark this answers is not in the document]
Comment (Uri-Bar) [reply packages/salesforce-adapter/src/change_validators/active_flows.ts:193]: no, we have different CV for that, we cant delete any flows

File: packages/salesforce-adapter/src/change_validators/active_flows.ts [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    boolean => (
+  change.data.before.value.status === ACTIVE && change.data.after.value.status !== ACTIVE
+)
+
+const isActivatingChange = (change: ModificationChange<InstanceElement>):
+    boolean => (
+  change.data.before.value.status !== ACTIVE && change.data.after.value.status === ACTIVE
+)
+
+const isStatusChangeOnly = (change: ModificationChange<InstanceElement>):
```
[the remark this answers is not in the document]
Comment (Uri-Bar) [reply packages/salesforce-adapter/src/change_validators/active_flows.ts:55]: thats right

File: packages/salesforce-adapter/src/change_validators/active_flows.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (amitgoldarb) [reply packages/salesforce-adapter/src/change_validators/active_flows.ts:175]: Looking at the other CV, I think we can merge into this one. WDYT?

File: packages/salesforce-adapter/src/change_validators/active_flows.ts [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+}
+
+/**
+ * Handling addition and modification regarding active flows
+ */
+const activeFlowValidator = (config: SalesforceConfig, isSandbox: boolean): ChangeValidator =>
+  async (changes, elementsSource) => {
+    const flowSettings = isUndefined(elementsSource)
+      ? undefined : await elementsSource.get(new ElemID(SALESFORCE, 'FlowSettings', 'instance'))
+    const isPreferActiveVersion = config.fetch?.preferActiveFlowVersions ?? false
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/change_validators/active_flows.ts:170]: I'm wondering if we should extract this default value to a const and reuse else where. I'm speaking specifically of this false

File: packages/salesforce-adapter/src/change_validators/active_flows.ts [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    const isEnableFlowDeployAsActiveEnabled = isUndefined(flowSettings)
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/change_validators/active_flows.ts:171]: We don't type-gaurd this value. Should we add a typegaurd here? What would the default value be if this settings doesn't exist (I see we already chose false we should stick with that).

**Review by amitgoldarb — COMMENTED** [review]
Looking good!
Added a few mostly small comments.

File: packages/salesforce-adapter/src/change_validators/active_flows.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Uri-Bar) [reply packages/salesforce-adapter/src/change_validators/active_flows.ts:175]: possible, will do

**Review by amitgoldarb — APPROVED** [review]
LGTM!
Let's double check the strings with @tomermevorach before merging.
Great work!

File: packages/salesforce-adapter/src/change_validators/flows.ts [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+      message: 'Deploying these changes will create a new active version of this flow',
+      detailedMessage: `Deploying these changes will create a new active version of this flow in case the test coverage percentage is greater than the number specified in your salesforce org config. Otherwise, a new inactive version of this flow will be created. Flow name: ${instance.elemID.getFullName()}`,
+      deployActions: testCoveragePostDeploy(),
+    }
+  }
+  return {
+    elemID: instance.elemID,
+    severity: 'Info',
+    message: 'Your salesforce org is configured to disallow modifications to active flows',
+    detailedMessage: `Your salesforce org is configured to disallow modifications to active flows. Therefore, deploying these changes will create a new inactive version that will need to be activated manually.. Flow name: ${instance.elemID.getFullName()}`,
```
Comment (tomermevorach) [inline packages/salesforce-adapter/src/change_validators/flows.ts:128]: I think there is a redundant dot in this sentense

File: packages/salesforce-adapter/src/change_validators/flows.ts [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+      severity: 'Info',
+      message: 'Deploying these changes will create a new inactive version of this flow',
+      detailedMessage: `Deploying these changes will create a new inactive version of this flow. Flow name: ${instance.elemID.getFullName()}`,
+    }
+  }
+  return {
+    elemID: instance.elemID,
+    severity: 'Info',
+    message: ' Deploying these changes will create a new inactive version of this flow',
+    detailedMessage: `Bear in mind that the new inactive version will not appear in Salto since your Salto environment is configured to prefer fetching active flow versions. Flow name: ${instance.elemID.getFullName()}`,
```
Comment (tomermevorach) [inline packages/salesforce-adapter/src/change_validators/flows.ts:93]: I think this version of text should be returned in the if(preferActive) section

**Pull request merged**